util/reggen/fpv_csr.sv.tpl

```diff
@@ -56,33 +68,47 @@ module ${mod_base}_csr_assert_fpv import tlul_pkg::*;
   assign a_mask_bit[23:16] = h2d.a_mask[2] ? '1 : '0;
   assign a_mask_bit[31:24] = h2d.a_mask[3] ? '1 : '0;
 
+  bit [${addr_msb}-2:0] hro_idx; // index for exp_vals
+  bit [${addr_msb}:0]   normalized_addr;
+
+  // Map register address with hro_idx in exp_vals array.
+  always_comb begin: decode_hro_addr_to_idx
+    unique case (pend_trans[d2h.d_source].addr)
+% for hro_reg in hro_regs_list:
+      ${hro_reg.offset}: hro_idx <= ${hro_idx};
 <%
-  addr_width = rb.get_addr_width()
-  addr_msb  = addr_width - 1
-  block_size = ((rb.flat_regs[-1].offset) >> 2) + 1
+  hro_map[hro_reg.offset] = hro_idx;
```

Comment:
I'd probably suggest constructing `hro_map` at the top of the file, rather than doing it here. Otherwise, this stateful computation is mixed in with the template expansion, which is probably a bit confusing to readers.

Reply:
I think that `hro_map` is still being constructed here? I was imagining something like
```python3
hro_map = {r.offset: idx for idx, r in enumerate(hro_regs_list)}
```
at the top. Or, maybe better:
```python3
hro_map = {r.offset: (idx, r) for idx, r in enumerate(hro_regs_list)}
```
and then iterate down here with `for idx, r in hro_map.values()` (and not use `hro_regs_list` after we've constructed `hro_map`)?

I've not tested this, so sorry if I'm talking rubbish!

Reply:
Oh thanks Rupert for your explanation! Sorry I did not fully get your suggestion. I see I see! That is very useful suggestions.. Let me give that a try.